### ISSUE TITLE
Support proposed additional export ___ from "mod" statements.

### ISSUE THREAD

**Issue by leebyron** [babel/babel#1091]
In March 2015 TC39 meeting, this proposal was discussed and pretty unanimously supported:
https://github.com/leebyron/ecmascript-more-export-from

The suggested next steps were:

* Write up full spec text
* Add to Babel transpiler (as experimental feature) to start getting feedback quickly.

The plan is to advance the proposal through the stages offline, getting an official Stage 2 or 3 at the next meeting.

I'm happy to write a PR to add this, but would also be happy for anyone else to take this on if motivated.

**Comment by sebmck** [comment]
I'm happy to take this on if you haven't started already.

**Comment by leebyron** [comment]
I haven't no. Feel free!

**Comment by leebyron** [comment]
I added spec in the proposal repo that describes the grammar

**Comment by sebmck** [comment]
How about the following ESTree definitions?

#### `export * as foo from "bar";`

```javascript
extend interface ExportAllDeclaration {
  exported: Identifier;
}
```

#### `export foo from "bar";`

```
interface ExportDefaultSpecifier {
    type: "ExportDefaultSpecifier";
    exported: Identifier;
}
```

**Comment by sebmck** [comment]
Actually, having `export foo from "bar";` be a `ExportNamedDeclaration` is pretty dirty. `ExportDefaultDeclaration` is already taken so what about `ExportNamespaceDeclaration`?

```
interface ExportNamespaceDeclaration {
  type: "ExportNamespaceDeclaration";
  exported: Identifier;
  source: Literal;
}
```

**Comment by sebmck** [comment]
Just went with the AST definitions: 

```
interface ExportNamespaceDeclaration {
  type: "ExportNamespaceDeclaration";
  exported: Identifier;
  source: Literal;
}

extend interface ExportAllDeclaration {
  exported: Identifier;
}
```

Feel free to suggest alterations! I'm not too confident in them.

**Comment by leebyron** [comment]
I would expect anything with "Namespace" to refer to the `export * as ns from "mod"` syntax. And the extentsion to `ExportAllDeclaration` would mean that it has very different semantic meaning whether or not exported is non-null, which probably isn't great.

I'm reading https://github.com/estree/estree/blob/master/es6.md#importdeclaration to try to find the direct parallel, but the ESTree spec looks under-specified. Not sure when `ImportDefaultSpecifier` is referenced, for example. Also, unclear how `import a, {b as x, c} from "mod"` or `import a, * as ns from "mod"` should be represented by ESTree.

The goal of my proposal is symmetry between `export ___ from` and `import ___ from`, so presumably the AST should follow the same symmetry:

My suggestion:

```diff
+interface ExportFromDeclaration <: Node {
+    type: "ExportFromDeclaration";
+    specifiers: [ ExportFromSpecifier ];
+    source: Literal;
+}
+
+interface ExportFromSpecifier {}
+
+interface ExportAllFromSpecifier <: Node, ExportFromSpecifier {
+    type: "ExportAllFromSpecifier";
+}
+
+interface ExportDefaultFromSpecifier <: Node, ExportFromSpecifier {
+    type: "ExportDefaultFromSpecifier";
+    exported: Identifier;
+}
+
+interface ExportNamedFromSpecifier <: Node, ExportFromSpecifier {
+    type: "ExportNamedFromSpecifier";
+    imported: Identifier;
+    exported: Identifier;
+}
+
+interface ExportNamespaceFromSpecifier <: Node, ExportFromSpecifier {
+    type: "ExportNamespaceFromSpecifier";
+    exported: Identifier;
+}
+

 interface ExportNamedDeclaration <: Node {
     type: "ExportNamedDeclaration";
     declaration: Declaration | null;
     specifiers: [ ExportSpecifier ];
-    source: Literal | null;
 }

-interface ExportAllDeclaration <: Node {
-    type: "ExportAllDeclaration";
-    source: Literal;
-}
```

** export a, {b as x, c} from "mod" **

```js
{ type: "ExportFromDeclaration"
  specifiers: 
    [ { type: "ExportDefaultFromSpecifier"
        exported: { type: "Identifier", name: "a" } },
      { type: "ExportNamedFromSpecifier"
        imported: { type: "Identifier", name: "b" }
        exported: { type: "Identifier", name: "x" } },
      { type: "ExportNamedFromSpecifier"
        imported: { type: "Identifier", name: "c" }
        exported: { type: "Identifier", name: "c" } } ]
  source: { type: "Literal", value: "mod" } }
```

**Comment by sebmck** [comment]
@leebyron Yeah, it is very underspecified. It's mainly meant as a guide for parser maintainers and tooling that's already familiar with the current SpiderMonkey AST. Non-additive changes are out of the question for the AST definitions. `ImportDefaultSpecifier` is the `foo` in `import foo from "bar";`. While `import a, * as ns from "mod"` would be interpreted as an `ImportDeclaration` with the specifiers of `ImportDefaultSpecifier` and `ImportNamespaceSpecifier`.

**Comment by leebyron** [comment]
Non-additive still out of the question for deltas to ES6?

But note that this is still actually purely additive. The removed `source` field from `ExportNamedDeclaration` may just be always null after this proposal, and `ExportAllDeclaration` simply never used.

**Comment by sebmck** [comment]
> Non-additive still out of the question for deltas to ES6?

Yep.

> But note that this is still actually purely additive.

`ExportAllDeclaration` is removed which is a non-additive change.

**Comment by leebyron** [comment]
Not sure if ESTree likes to support unions within an array, but if so:

```diff
+interface ExportFromDeclaration <: Node {
+    type: "ExportFromDeclaration";
+    specifiers: [ ExportFromSpecifier | ExportDefaultFromSpecifier | ExportNamespaceFromSpecifier | ExportAllFromSpecifier ];
+    source: Literal;
+}
+
+interface ExportFromSpecifier <: Node {
+    type: "ExportFromSpecifier";
+    imported: Identifier;
+    exported: Identifier;
+}
+
+interface ExportDefaultFromSpecifier <: Node {
+    type: "ExportDefaultFromSpecifier";
+    exported: Identifier;
+}
+
+interface ExportNamespaceFromSpecifier <: Node {
+    type: "ExportNamespaceFromSpecifier";
+    exported: Identifier;
+}
+
+interface ExportAllFromSpecifier <: Node {
+    type: "ExportAllFromSpecifier";
+}
```

**Comment by leebyron** [comment]
> ExportAllDeclaration is removed which is a non-additive change.

Assume that rather than removing it, I simply suggest that when this proposal spec is enabled, it's just never used by the parser.

**Comment by leebyron** [comment]
You could certainly also leave `ExportAllDeclaration` alone completely - it's still used for its purpose, but proposed addition is:

```
+interface ExportFromDeclaration <: Node {
+    type: "ExportFromDeclaration";
+    specifiers: [ ExportFromSpecifier | ExportDefaultFromSpecifier | ExportNamespaceFromSpecifier ];
+    source: Literal;
+}
+
+interface ExportFromSpecifier <: Node {
+    type: "ExportFromSpecifier";
+    imported: Identifier;
+    exported: Identifier;
+}
+
+interface ExportDefaultFromSpecifier <: Node {
+    type: "ExportDefaultFromSpecifier";
+    exported: Identifier;
+}
+
+interface ExportNamespaceFromSpecifier <: Node {
+    type: "ExportNamespaceFromSpecifier";
+    exported: Identifier;
+}
```

**Comment by leebyron** [comment]
^^ That's probably better anyway, because the "ExportAllFromSpecifier" would never be legally combined with any other export-from specifier, so it's pretty logical to have it stand alone.

**Comment by sebmck** [comment]
Looks good to me but what's `ExportFromSpecifier`?

**Comment by leebyron** [comment]
```js
export a, {b as x, c} from "mod";
```

Parses to:

```js
{ type: "ExportFromDeclaration"
  specifiers: 
    [ { type: "ExportDefaultFromSpecifier"
        exported: { type: "Identifier", name: "a" } },
      { type: "ExportFromSpecifier"
        imported: { type: "Identifier", name: "b" }
        exported: { type: "Identifier", name: "x" } },
      { type: "ExportFromSpecifier"
        imported: { type: "Identifier", name: "c" }
        exported: { type: "Identifier", name: "c" } } ]
  source: { type: "Literal", value: "mod" } }
```

**Comment by leebyron** [comment]
I suppose you could re-use `ExportSpecifier` in place of `ExportFromSpecifier`. The only real difference is whether you refer to the imported identifier as `imported` or `local`.

**Comment by sebmck** [comment]
That's currently represented by `ExportSpecifier` and `ExportNamedDeclaration` so it can't be changed.

**Comment by leebyron** [comment]
Not recommending changing it, just using it. So `ExportFromDeclaration` just also refers to `ExportSpecifier`

**Comment by leebyron** [comment]
```js
interface ExportFromDeclaration <: Node {
    type: "ExportFromDeclaration";
    specifiers: [ ExportSpecifier | ExportDefaultSpecifier | ExportNamespaceSpecifier ];
    source: Literal;
}

interface ExportDefaultSpecifier <: Node {
    type: "ExportDefaultFromSpecifier";
    exported: Identifier;
}

interface ExportNamespaceSpecifier <: Node {
    type: "ExportNamespaceFromSpecifier";
    exported: Identifier;
}
```

**Comment by leebyron** [comment]
In which case:

```js
export a, {b as x, c} from "mod";
```

Parses to:

```js
{ type: "ExportFromDeclaration"
  specifiers: 
    [ { type: "ExportDefaultFromSpecifier"
        exported: { type: "Identifier", name: "a" } },
      { type: "ExportSpecifier"
        exported: { type: "Identifier", name: "x" },
        local: { type: "Identifier", name: "b" } },
      { type: "ExportSpecifier"
        exported: { type: "Identifier", name: "c" }, 
        imported: { type: "Identifier", name: "c" } } ],
  source: { type: "Literal", value: "mod" } }
```

**Comment by sebmck** [comment]
So the idea is for `export {b as x, c} from "mod";` turns from a `ExportDeclaration` into a `ExportFromDeclaration` when it's changed to `export a, {b as x, c} from "mod";`? If so, that seems rather unintuitive and I'd imagine it'd be unlikely to be accepted if proposed.

**Comment by leebyron** [comment]
Not exactly what I'm proposing. I propose that when this spec is also being considered by a parser, that `export {b as x, c} from "mod";` should always result in `ExportFromDeclaration`.

**Comment by sebmck** [comment]
That's even more unlikely to be considered. It breaks backwards compatibility with the AST as once it's accepted into ES tooling will have to return `ExportFromDeclaration` always.

**Comment by leebyron** [comment]
Hmm, I'm at a loss then. Shimming this in is going to be awkward no matter how you look at it if a parser must parse both valid ES6 and ES7 without additional information.

**Comment by leebyron** [comment]
> So the idea is for export {b as x, c} from "mod"; turns from a ExportDeclaration into a ExportFromDeclaration when it's changed to export a, {b as x, c} from "mod";?

This is the best I can think up at the moment then to preserve the ability to parse valid ES6 without additional information.

**Comment by sebmck** [comment]
Yeah, it's extremely finnicky. cc @caridy

**Comment by leebyron** [comment]
Ok, lets back up to your original proposal, @sebmck. Here's a version of that which solves my concerns about "namespace" being confusing and altering "export-all" changing semantics, but also solves your concerns about being purely additive:

```js
extend interface ExportNamedDeclaration <: Node {
    specifiers: [ ExportDefaultSpecifier | ExportNamespaceSpecifier | ExportSpecifier ];
}

interface ExportDefaultSpecifier <: Node {
    type: "ExportDefaultSpecifier";
    exported: Identifier;
}

interface ExportNamespaceSpecifier <: Node {
    type: "ExportNamespaceSpecifier";
    exported: Identifier;
}
```

In this case `export {b as x, c} from "mod";` and `export a, {b as x, c} from "mod";` both parse as `ExportNamedDeclaration`, but now we have the additional two interfaces to support these two new specifier forms.

**Comment by leebyron** [comment]
In which case:

```js
export * from "mod";
```

Is still "ExportAllDeclaration" (unchanged).

And

```js
export * as ns from "mod";
```

Parses to:

```js
{ type: "ExportNamedDeclaration"
  specifiers: 
    [ { type: "ExportNamespaceSpecifier"
        exported: { type: "Identifier", name: "ns" } } ],
  source: { type: "Literal", value: "mod" } }
```

And


```js
export a, {b as x, c} from "mod";
```

Parses to:

```js
{ type: "ExportNamedDeclaration"
  specifiers: 
    [ { type: "ExportDefaultSpecifier"
        exported: { type: "Identifier", name: "a" } },
      { type: "ExportSpecifier"
        exported: { type: "Identifier", name: "x" },
        local: { type: "Identifier", name: "b" } },
      { type: "ExportSpecifier"
        exported: { type: "Identifier", name: "c" }, 
        imported: { type: "Identifier", name: "c" } } ],
  source: { type: "Literal", value: "mod" } }
```

Much like how: 

```js
export {b as x, c} from "mod";
```

Currently parses to:

```js
{ type: "ExportNamedDeclaration"
  specifiers: 
    [ { type: "ExportSpecifier"
        exported: { type: "Identifier", name: "x" },
        local: { type: "Identifier", name: "b" } },
      { type: "ExportSpecifier"
        exported: { type: "Identifier", name: "c" }, 
        imported: { type: "Identifier", name: "c" } } ],
  source: { type: "Literal", value: "mod" } }
```

**Comment by sebmck** [comment]
That looks perfect to me. Probably as good as you can get it.

**Comment by leebyron** [comment]
Sweet. I've added this to https://github.com/leebyron/ecmascript-more-export-from/blob/master/ESTree.md so it's easier for others to find and read :)

**Comment by sebmck** [comment]
Available as of 5.0.0: http://babeljs.io/blog/2015/03/31/5.0.0/#stage-1:-export-extensions
